src/Features/Core/Portable/FindUsages/SourceReferenceItem.cs

```diff
@@ -25,11 +34,49 @@ internal sealed class SourceReferenceItem
         /// </summary>
         public bool IsWrittenTo { get; }
 
+        /// <summary>
+        /// Additional information about the reference.
+        /// Each entry represents a key-values pair of data. For example, consider the below entry:
+        ///     { "ValueUsageInfo" } = { "Read", "Write" }
+        /// This entry indicates that the reference has additional value usage information which indicate
+        /// it is a read/write reference, such as say 'a++'.
+        /// </summary>
+        public ReferenceInfoMap ReferenceInfo { get; }
+
+        [Obsolete]
         public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, bool isWrittenTo)
         {
             Definition = definition;
             SourceSpan = sourceSpan;
             IsWrittenTo = isWrittenTo;
+            ReferenceInfo = ReferenceInfoMap.Empty;
+        }
+
+        public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, ReferenceInfoMap referenceInfo)
+        {
+            Definition = definition;
+            SourceSpan = sourceSpan;
+            ReferenceInfo = referenceInfo ?? ReferenceInfoMap.Empty;
+        }
+
+        internal SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, ValueUsageInfo valueUsageInfo)
+            : this(definition, sourceSpan, GetOrCreateReferenceInfo(valueUsageInfo))
+        {
+            IsWrittenTo = valueUsageInfo.IsWrittenTo();
+        }
+
+        private static ReferenceInfoMap GetOrCreateReferenceInfo(ValueUsageInfo valueUsageInfo)
+            => s_valueUsageInfoToReferenceInfoMap.GetOrAdd(valueUsageInfo, CreateReferenceInfo);
```

Comment:
i can't remember.  will this cache the dleegate for CreateReferenceInfo?  I don't think it will.  So we can either fix this to explicitly cache that delegate, or switch to a simple lambda that calls the function (since hte compiler will cache htat).